packages/labs/forms/src/form-associated.ts

```diff
@@ -0,0 +1,227 @@
+
+  role: ElementInternals['role'];
+
+  new (): FormAssociated;
+}
+
+interface ValidityResult {
+  flags: ValidityState;
+  message?: string;
+  anchor?: HTMLElement;
+}
+
+const internalsCache = new WeakMap<FormAssociated, ElementInternals>();
+
+/**
+ * Returns the ElementInternals for a FormAssociated element. This should only
+ * be called by subclasses of FormAssociated.
+ */
+export const getInternals = (element: FormAssociated) =>
+  internalsCache.get(element);
+
+/**
+ * Returns true if the element is disabled.
+ */
+export const isDisabled = (element: FormAssociated) =>
+  disabledElements.has(element);
+
+const disabledElements = new WeakSet<FormAssociated>();
+
+/**
+ * A mixin that makes a ReactiveElement into a form-associated custom element.
+ *
+ * - ARIA role: Sets this element's ElementInternals role to the value of the
+ *   static `role` property.
+ * - Form value: The value of the element is stored in the field decorated with
+ *   the `@formValue()` decorator. This value field can have any name, but the
+ *   type must be assignable to `string | File | FormData | null`.
+ * - Form state: The state of the element is stored in the field decorated with
+ *   the `@formState()` decorator. This state field can have any name, but the
+ *   type must be assignable to `string | File | FormData | null`. The state
+ *   field should be private, as it is only intended to be set by the
+ *   FormAssociated mixin. The state field should usually be a getter/setter
+ *   pair so that the form value can be derived from the state during form
+ *   state restoration.
+ * - Form reset: When the form is reset, the value field is set to its initial
+ *   value, and the state field is set to its initial state. This means that the
+ *   initial value and state will be stored for the lifetime of the element.
+ * - Form state restoration: When the form is restored, the mode can be either
+ *   'restore' or 'autocomplete'. In 'restore' mode, the state field is set to
+ *   the state setter, if present, is set to the state that was previously
+ *   passed to `internals.setFormValue()`. It's the state setter's
+ *   responsibility to update the value setter. If there is no state setter, the
+ *   value setter is updated directly. In 'autocomplete' mode, the value setter
+ *   is updated directly.
+ * - Form disabled: When the element is disabled, the `ariaDisabled` property of
+ *   the ElementInternals is set to "true". The element is also added to a set
+ *   of disabled elements, which can be checked with the `isDisabled()`
+ *   function.
+ */
+export const FormAssociated = <T extends Constructor<ReactiveElement>>(
+  base: T
+) => {
+  class C extends base {
+    #internals: ElementInternals;
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    constructor(...args: any[]) {
+      super(...args);
+      const internals = this.attachInternals();
+      this.#internals = internals;
+      internalsCache.set(this, internals);
```

Comment:
This strikes me as a bit cumbersome v. just using a symbol based property and sharing that symbol (possibly even putting it on the constructor).

Reply:
I was hoping to find a way to limit the call so that it could only be called once per subclass, like `attachInternals()`. Then if a subclass called it and stored the value, a user couldn't call it later and mess with the element.

Reply:
Sorry, not following. Limit the call to what, the property?

Reply:
Limit it to being called just once. `attachInternals()` throws if called a separate time, to help prevent outside access to it. An element can either call `attachInternals()` in the constructor (or set a static on the class, but I forget the name. Something like `static disableInternals = true`) to prevent anyone else from getting to internals.

I'd like to do the same for `getInternals()` so it doesn't bypass those protections.

Reply:
`disabledFeatures`. It's very poorly documented.

Reply:
I updated `getInternals()` to require a constructor that's on the proto chain be passed in, and allows only one call per proto. That can be worked around, but hopefully this prevents casual misuse.